binstring/binstring.ml

```diff
@@ -0,0 +1,125 @@
+module Str = Stdlib.String
+module Buf = Stdlib.Buffer
+
+type t = { inner : string list; offset : int; length : int }
+
+let empty = { inner = []; length = 0; offset = 0 }
+let length t = t.length
+
+exception Malformed of string
+
+let of_string str =
+  let length = Str.length str in
+  let buf = Bytes.create length in
+  let decoder = Uutf.decoder ~encoding:`UTF_8 (`String str) in
+  let pos = ref 0 in
+  let add char =
+    let n = Bytes.set_utf_8_uchar buf !pos char in
```

Comment:
Is this setting an UTF-8 at any random position? That sounds expensive.

Reply:
we have a preallocated buffer here with the size of the string that we're verifying for utf-8 encoding, so this is more of an array.set.

in any case, after our chat, i'm removing this verification and leaving it to the extraction functions.